src/pepr/operator/controllers/keycloak/clients/client-credentials.ts

```diff
@@ -0,0 +1,143 @@
+const UDS_OPERATOR_CLIENT_ID = "uds-operator";
+let cachedToken: string | null = null;
+
+export interface KeycloakToken {
+  exp: number;
+  resource_access: {
+    "realm-management": {
+      roles: string[];
+    };
+    [key: string]: unknown;
+  };
+
+  [key: string]: unknown;
+}
+
+export function resetCachedToken() {
+  cachedToken = null;
+}
+
+export async function credentialsGetAccessToken() {
+  if (cachedToken) {
+    try {
+      const jwt = parseKeycloakToken(cachedToken);
+      if (jwt.exp && jwt.exp > Math.floor(Date.now() / 1000) + 5) return cachedToken;
+    } catch (e) {
+      log.error(e, "Failed to parse cached token");
+      cachedToken = null;
+    }
+  }
+
+  const secret = await K8s(kind.Secret).InNamespace(SECRET_NAMESPACE).Get(SECRET_NAME);
+  if (!secret) throw new Error("Missing secret");
+  const encodedSecret = secret.data?.[UDS_OPERATOR_CLIENT_ID];
+  if (!encodedSecret) throw new Error("Missing client secret");
+
+  const clientSecret = Buffer.from(encodedSecret, "base64").toString("utf-8");
+  const params = new URLSearchParams();
+  params.append("grant_type", "client_credentials");
+  params.append("client_id", UDS_OPERATOR_CLIENT_ID);
+  params.append("client_secret", clientSecret);
+
+  const response = await fetch<KeycloakAccessTokenResponse>(clientCredentialsUrl, {
+    method: "POST",
+    headers: { "Content-Type": "application/x-www-form-urlencoded" },
+    body: params.toString(),
+  });
+  await throwErrorIfNeeded(response);
+  cachedToken = response.data.access_token;
+  return cachedToken;
+}
+
+export async function credentialsCreateOrUpdate(client: Partial<Client>) {
+  log.info(`credentialsCreateOrUpdate: creating/updating client ${JSON.stringify(client)}`);
+  const existingClient = await credentialsGet(client);
+  if (existingClient) {
+    return credentialsUpdate(client);
+  } else {
+    return credentialsCreate(client);
+  }
+}
+
+export async function credentialsCreate(client: Partial<Client>) {
+  log.info(`credentialsCreate: creating client ${JSON.stringify(client)}`);
+  const token = await credentialsGetAccessToken();
+  const response = await fetch(clientsAdminUrl, {
+    method: "POST",
+    headers: { "Content-Type": "application/json", Authorization: `Bearer ${token}` },
+    body: JSON.stringify(client),
+  });
+  await throwErrorIfNeeded(response, () => {
+    resetCachedToken();
+  });
+  return credentialsGet(client);
+}
+
+export async function credentialsGet(client: Partial<Client>) {
+  log.info(`credentialsGet: retrieving client ${JSON.stringify(client)}`);
+  const token = await credentialsGetAccessToken();
+  const url = `${clientsAdminUrl}?clientId=${encodeURIComponent(client.clientId!)}`;
+  const response = await fetch<ClientWithId[]>(url, {
+    method: "GET",
+    headers: { "Content-Type": "application/json", Authorization: `Bearer ${token}` },
+  });
+  await throwErrorIfNeeded(response, () => {
+    resetCachedToken();
+  });
+  return response.data[0];
```

Comment:
Is this data field the client id, full client object, something else? Might be good to comment on this code to take note of what we are returning (mostly wondering because with the dynamic response we return the full data, I assume this is different but it's hard to tell what this response looks like).

Reply:
This is one of the quirks of the Keycloak Admin APIs.

There's no Client GET REST endpoint that obtains a client based on `client_id` (the logical client name, like `uds-operator`). All Admin REST endpoints for client operator on the database Client ID, which is a UUID. The only interface that allows to obtain the Client using the `client_id` is the collection interface which returns you the client in a singular collection. I know - it's a terrible design but we'll be addressing this in v2 Admin API initiative. 

I'll add a comment explaining this.

Reply:
Thanks this is perfect!